### DISCUSSION TITLE
Making of a Tutorial on SRM with losses.

### DISCUSSION THREAD

**Discussion by Ksp-3086** [Eomys/pyleecan#449] [General]
This discussion is a continuation for the #428  for the SRMs.

**Comment by Ksp-3086** [comment]
Hello @SebGue ,
You mentioned that you put T=1/p th of a rotation. Is it the rotation in one second or one minute as it depends on the rpm set for the machine. If you are referring to the rotation in one second than I am thinking the formula must be T=60/(p x N). You also mentioned a phase shift, Can I ask how much? How did you reach your conclusion for a phase shift.

So the formula for tau is dependent on T that is fine by me as it was before. When I sent the previous file, which motor did you simulate, the 3 phase one or the 4 phase one?

**Comment by SebGue** [comment]
Yes, since time is in seconds and electrical period depends on rotational speed it is T = 60 / p / N
The phase shift can be be estimated by the DC current test i ran earlier. I set it to 1/8 of T. So you got to do something like `I1 = rectwave(time-1/8)`
You also got to consider the order you drive your phases in relation to the rotational direction if you haven't done yet. I.e. my phase order was A D C B.

**Comment by Ksp-3086** [comment]
So I put the following changes `T_P=60/(n*N)` and `tau=T_P/qs` to the code as you said. For the time I put, `time1=(linspace(0,T_P,32))-(T_P/8) ` and also `I_1=rect_wav(A,time1-1/8,T_P,tau)` at different times. I have also changed the phase sequence but still I am getting unfavorable results. You can take a look at the code if you like.
[SRM_innerrotor1.zip](https://github.com/Eomys/pyleecan/files/7278402/SRM_innerrotor1.zip)

**Comment by SebGue** [comment]
You don't have to shift change your 'time1'. Only shift the input of rect_wav as above. 
But you should take care for your rect_wav function, since it only works for times between 0 and T.

**Comment by Ksp-3086** [comment]
You mean like `I_1=rect_wav(A,time1-1/8,T_P,tau)` I got a straight line of torque actually. I even tried `I_1=rect_wav(A,time1-T_P/8,T_P,tau)` still unfavorable results.

**Comment by SebGue** [comment]
...sorry, I missed that you substract 1/8 but you got to substract 1/8*T_P.

**Comment by Ksp-3086** [comment]
Yeah I did that too , right now I am using the code `I_1=rect_wav(A,time1-1/8,T_P-1/8,tau)` so that it works between 1/8 and T_P supposedly. Still I am getting a flat torque graph due to no current.

**Comment by SebGue** [comment]
`I_1=rect_wav(A,time1-T_P/8,T_P,tau)` should work. Whats your current vector/matrix then.

**Reply by Ksp-3086** [reply]
![Screenshot 2021-10-04 202007](https://user-images.githubusercontent.com/84498303/135873430-4e2900e1-5d46-4db2-918c-3201017619fb.png)
This is the current matrix and the torque curve is
![Figure_2](https://user-images.githubusercontent.com/84498303/135873496-3a85d7d9-d5a7-4e23-8a7d-2ab4945ca72c.png)

**Comment by SebGue** [comment]
With `T_P=60/(n*N)` n has to be the number of rotor poles, not the stator poles.

**Comment by SebGue** [comment]
... finally you have should set linspace's endpoint to false

**Reply by Ksp-3086** [reply]
So I did as you said so the current is still the same, the graph comes out as
![Figure_2, rotor poles p](https://user-images.githubusercontent.com/84498303/135877303-949e9062-1ec5-4a7a-a81c-ee0fab71e8fe.png)

**Reply by SebGue** [reply]
There is something wrong. The period should only be 10ms @ 1000rpm and 6 rotor poles.

**Reply by Ksp-3086** [reply]
Ok sorry, when I was experimenting with different poles positions, I changed it to 4 poles. Now still I am getting this.

![Figure_2, going to 6 poles](https://user-images.githubusercontent.com/84498303/135890704-00f3dc98-b167-4546-917d-d52424bc671a.png)

**Reply by SebGue** [reply]
But now it looks ok. The ripple left, is due to sub optimal phase shift.
The zero torque at the beginning is due to your rect wave. It doesn't repeat the sequence (also for neqative time).

**Reply by Ksp-3086** [reply]
So.. anything else that I am missing that I should do?

**Reply by Ksp-3086** [reply]
So, the phase shift is somewhat not ok, if I got that right? Did I put the phase shift wrong or is there something else. I understand now how the time period of the wave is taken. I was going initially according to the rotation of the stator while I should have been taking for the rotor. Also, when the machine gets plotted while running the code, I did notice that phase sequence A-B-C-D was in anticlockwise direction while the motor goes in clockwise direction.

You did ask what did I mean by the back emf , right? I am working on making an efficiency map for SRM which is when I noticed that experimentally the map is similar in shape to DC motors. The reason I think you may know is due to high back emf at high speeds so at high rpm only lower current is used. So in the simulation at high rpm such as 3000, shouldn't the torque be less due to back emf?

**Reply by SebGue** [reply]
The phase shift seems not optimal. But I can't tell you the right value, since it depends on your drive system.

As you may know, torque in most cases is proprtional to current. So if you imply current, there is no reduction in torque no matter what speed.
Back emf (don't know if its called that way in SRM) is all about voltage.

**Reply by Ksp-3086** [reply]
I mean I am using the same file which you showed the results for, 6 pole rotor and 8 pole stator with 4 phases. I have not chosen a drive system properly here so I am not sure how to get a phase shift.

About the back emf, is there a way to simulate it at high speeds or something? I mean yeah it does depend on applied voltage so I may need to change the whole way that the input is being given. Right now according to the input to be given, the parameters are current and rpm. Is there a way to change it so that we could get a back emf?

**Reply by SebGue** [reply]
I set my phase shift by 'hacking' your rect_wav function. but that's quick and dirty and shouldn't be done for reproductive results. So the 1/8th was only a rough estimate.
To get an appropriate phase shift you can do the DC test (no matter what actual machine) for some fraction of your period time. Then you will see, when torque is at its max. and when phase A (B, C, ...) sequence should be started.

For the emf (also known as induced voltage) you may use the calculated value in OutMag.emf. It is basically the time derivative of phase flux linkage (i.e. the magnetic flux linked to the phase coils times the coils number of turns).
For further explainations, I don't have sufficient time to spend :-/

BTW. What's your actual focus as a mechanical engineer?

**Reply by Ksp-3086** [reply]
So, I made a design for an in wheel motor or a hub motor for automobiles starting from motorbikes first. Although, it is supposed to have better feedback and control over torque and rpm using vehicle control methods, the starting basis is the motor. It was something I though of and my Professor told me to get an efficiency map to get the funding started from investors. I did not have enough data and figures just the design and working principles. You could say the main focus is as an Automotive Engineer as I am doing my masters in that field.

The DC test you are referring to, do you mean I should just test different phase shifts and see proper results which one I get? The graph you showed is completely different than mine which also looks more favorable. So, using the average of `out_femm.mag.emf`, I can determine if the resulting simulation is favorable or not? I am guessing using V=IR, R is resistance of the wires, the input voltage can be found. If the induced is higher then the simulation must give an error. Did I get that right?

**Reply by SebGue** [reply]
The DC test is just the thing with the DC current on one phase (preferably phase A) as done before.

As you can model mechanical systems by arrangements of springs, masses, dampers, ..., you can model an electrical system similarly with resistances, inductors, voltage sources, ... Thats an EEC and the number and arrangment of the elements determines the quality of your model. 
The simples model you can setup is a resistance (to model winding losses) in series with a voltage source (i.e. the induced voltage in the phase coil, the emf). So your voltage equation will be Va = R * Ia + EMFa with Va ... voltage in phase A, Ia ... current in phase A, EMFa ...

But you may not need that now, if you just want the efficiency map, since you only need mech. power + losses for that.

**Reply by Ksp-3086** [reply]
![Screenshot 2021-10-05 213155](https://user-images.githubusercontent.com/84498303/136060067-0edfe691-12ac-4e8c-aeab-3f94a8e30471.png)
I got this from a research paper, although the motor dimensions are different than the one I am using, you can see at high rpm there is no high torque but that is not the case here as I get the same amount of torque at every rpm. That is where this concept comes into play

**Reply by SebGue** [reply]
I know these figures, since they are normal to electrical machines and I also know you also want to have such figures for your machine.
But you got to understand that for your kind of simulation, no matter what the speed is, the torque will always depend on current. So there is no reduction in torque at higher speeds. 
You got to do that on your own at higher speeds. The reason that there is a reduction in torque is due to machines thermal restrictions, limited drive voltage and/or current, ...
So I can't really help you with that.

**Comment by Ksp-3086** [comment]
Hi @SebGue ,
Is it a bug or something, like one of my files which I have run a many times before, I tried changing the number of FEMM instances to run at the same time to 1 (just to check the outputs). It took a lot of time so I reverted back to 4. Now, whenever I try to run the file it takes a lot of time to run. Previously it took less than 20 minutes but now its taking more than an hour. All the other files are running just fine. I tried closing VS code and opening it too. The problem remains the same. Any fix?

**Reply by SebGue** [reply]
Don't know whats wrong. You can check the actual number of threads by counting the number of femm instances at a time.

**Reply by Ksp-3086** [reply]
Wait.. I know what is wrong. The code in which you asked me to do `time-T/8` (for phase shift), if I remove that part it runs. The FEMM instances otherwise are not able to reach a solution. It was working fine yesterday morning. Do you think this problem might be temporary? How is it on your end?

**Reply by SebGue** [reply]
Sorry, I don't know. My simulations run fine. Have you updated pyleecan recently? There were some changes. Maybe there is an issue with the update?

**Reply by Ksp-3086** [reply]
I have not updated anything but today morning I did update the VS code. It did not seem to work though. It looks more like a problem on my end. One of the FEMM instances does not produce a solution. I noticed when I opened them up in the background.

**Reply by Ksp-3086** [reply]
I opened up the FEMM directly and ran the simulation by opening the file, using options on it. Then I ran the script again, looks like it is working fine now. The problem was at the FEMM end.

**Comment by Ksp-3086** [comment]
Hi @SebGue ,
I got the torque as you are seeing right now.
![Untitled](https://user-images.githubusercontent.com/84498303/137725097-579c9eec-726b-4a7c-aaf9-550dbafc3aa9.png)
I have sent you the file. Can you take a look at it. Also it is taking abnormally very long to run like 6 hours or something.
[SRM_A.zip](https://github.com/Eomys/pyleecan/files/7364549/SRM_A.zip)

**Comment by SebGue** [comment]
Hello @Ksp-3086 

sorry was really busy for some time. Have you made any progress on your issue?
Otherwise I will have a look at your code this week.

Best regards, Sebastian

**Reply by Ksp-3086** [reply]
My internship is finished and I may not be able to work on the code anymore. I would like to however hand over the code for future references and future help. I also have some suggestions on the GUI but I think it would be better after I am able to do some experimental analysis.

**Reply by Ksp-3086** [reply]
So here is the code, I hope it could be of some help.
[SRM_innerrotor1.zip](https://github.com/Eomys/pyleecan/files/7475936/SRM_innerrotor1.zip)
